### PR TITLE
Fixes #18957 - move notifications to redux

### PR DESCRIPTION
**Pull request by matanwerbner**

### CODE REVIEW

File: app/views/home/_user_dropdown.html.erb [hunk 1]
```diff
@@ -13,4 +13,5 @@
   </ul>
 </li>
 
-<%= mount_react_component('NotificationDrawerToggle', '#notification_icon', {:url => notification_recipients_path}.to_json ) %>
\ No newline at end of file
+  <%= mount_react_component('NotificationDrawer','#notifications_container',{:url => notification_recipients_path}.to_json) %>
```
Comment (ohadlevy) [inline app/views/home/_user_dropdown.html.erb]: nitpick: extra whitespace.

File: webpack/assets/javascripts/foreman_tools.js [hunk 2]
```diff
@@ -44,6 +44,11 @@ export function activateTooltips(el = 'body') {
                                    }
                               });
   el.find('*[title]').not('*[rel]').tooltip({ container: 'body' });
+
+  // adding this to remove displayed tooltips for react-initiated
+  // calls to this function (no page:restore fired).
+  $('.tooltip.in').remove();
```
Comment (ohadlevy) [inline webpack/assets/javascripts/foreman_tools.js]: why is this here? shouldnt we avoid the in css class in the inigital tooltip initalization? e.g. https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/foreman_tools.js#L40

File: webpack/assets/javascripts/react_app/API.js [hunk 3]
```diff
@@ -1,7 +1,4 @@
-import ServerActions from './actions/ServerActions';
-import NotificationActions from './actions/NotificationActions';
-import { STATUS } from './constants';
-import $ from 'jquery';
+const $ = require('jquery');
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/API.js]: q: whats the diff between a const and import?

File: webpack/assets/javascripts/react_app/common/sessionStorage.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+if (!window.sessionStorage) {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/sessionStorage.js:1]: we used to do this with object.assign (see https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/bundle.js#L13). is this better approach then what we currently have?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js [hunk 5]
```diff
@@ -2,24 +2,22 @@ import React from 'react';
...
       <Dropdown.Toggle noCaret bsStyle="link">
         <Glyphicon bsClass="fa" glyph="ellipsis-v" />
       </Dropdown.Toggle>
       <Dropdown.Menu>
-        {listLinks}
+        {
+          links.map((link, i) =>
+            <MenuItem key={i}
+                      id={ `notification-kebab-${i}` }
+                      target={link.external ? '_blank' : '_self'}
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js]: note that @lizagilman also implemented this via detecting if the url is relative or not.

File: webpack/assets/javascripts/react_app/components/notifications/drawer/index.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        key={key}
+        onMarkAsRead={onMarkAsRead}
+        isExpanded={expandedGroup === key}
+        onExpand={onExpandGroup}
+        notifications={notificationGroups[key]}
+      />
+    ));
+  const noNotificationsMessage = (
+    <div id="no-notifications-container">
+      No Notifications...
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/index.js]: please extract strings.

File: webpack/assets/javascripts/react_app/components/notifications/drawer/index.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    </div>
+  );
+
+  return (
+    <div className="drawer-pf drawer-pf-notifications-non-clickable">
+      <div className="drawer-pf-title">
+        <h3 className="text-center">Notifications</h3>
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/index.js]: please extract.

**Review by ohadlevy — COMMENTED** [review]
thanks @matanwerbner , I've left a few comments/questions, also note the jest  tests are failing.

**Comment by matanwerbner** [comment]
@ohadlevy i can't see any jest failures, link?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js (hunk 5 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js]: i think external property is more clear, links can be relative or not, and open in window, or new window.

File: webpack/assets/javascripts/react_app/common/sessionStorage.js (hunk 4 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/common/sessionStorage.js:1]: It's better if you're looking to **add** properties to existing objects. in this case it is not needed. 
this line is used to mock sessionStorage in tests. i'm trying to find more elegant solutions to this.

File: webpack/assets/javascripts/foreman_tools.js (hunk 2 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/foreman_tools.js]: this is here because react components that change leave their opened tooltips open forever.
the solution: call initialize_tooltips whenever the component changes, and there remove the already opened tooltips.
also, i think it makes sense to remove existing tooltips when initializing.

**Comment by tbrisker** [comment]
I'd be happy to get rid of flux and move completely to redux, but as this stands, I cannot review this without spending several hours of understanding everything that changed and figuring out by myself which changes are actually related and which aren't. 
Please avoid changing directory structures this much in a PR that has another purpose. This makes the PR practically impossible to review as the combination of changes to files names and changes inside the files makes github miss the fact that a large part of the code didn't in fact change but rather moved from one place to another. I'm guessing out of the ~1.5K lines delta here only around 200-300 were actually related to the subject of the PR, and the others are just renaming stuff around, or unrelated refactors. If you want to refactor the structure or even the code in parts that aren't related, please do it in separate PRs, especially when the main change isn't trivial to understand.

**Comment by ohadlevy** [comment]
@matanwerbner please also remove the unneeded packages from package.json, i believe its both events and flux packages.

**Comment by ohadlevy** [comment]
also, /cc @gailsteiger

**Comment by matanwerbner** [comment]
@tbrisker if there is anything youre struggling witg, i will happily talk you through it. 
I agree that Redux is not very trivial, but if were going to use it, we may as well do it like proffesionals.

You can always checkout the branch, see notifications are working and that there are (finally) tests...

Bottom line, let me know if you need any explanations. I dont expect you to review the PR with no redux knowledge.

**Comment by tbrisker** [comment]
@matanwerbner I'm not talking about not understanding the redux code. I'm talking about renaming and moving dozens of files around that isn't related at all to the redux change, but makes review of this PR impossible, as well as quite a few other unrelated changes added inside this PR. 
A PR should contain only changes needed to address the issue it is meant to fix, not half a dozen other things you want to fix along the way, esp. when the main change itself is not trivial to understand.

File: webpack/assets/javascripts/react_app/redux/reducers/notifications/index.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import {
+  GET_NOTIFICATIONS,
+  SET_NOTIFICATIONS_DRAWER_OPENED,
+  SET_EXPANDED_GROUP,
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/reducers/notifications/index.js]: can you name this with a notification_prefix? its hard to filter the notifications actions with a page with many other events?

File: webpack/assets/javascripts/foreman_tools.js (hunk 2 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/foreman_tools.js]: @matanwerbner I'm not sure i follow your explanation? can you explain exactly the issue you faced and why this solves it? does it make sense in this case not to use bootstrap js and implement "react" one?

File: webpack/assets/javascripts/react_app/actions/NotificationActions.js [hunk 9]
```diff
@@ -1,39 +0,0 @@
-import API from '../API';
-import AppDispatcher from '../dispatcher';
-import NotificationsStore from '../stores/NotificationsStore';
-import { ACTIONS, STATUS } from '../constants';
-
-const TIMER = 10000;
-
-export default {
-  getNotifications(url) {
-    if ((document.visibilityState === 'visible' || document.visibilityState === 'prerender') &&
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/actions/NotificationActions.js:10]: I'm not seeing this condition in the new code? this will imply notification keeps on running regardless if the tab is in focus or not.

File: webpack/assets/javascripts/react_app/common/sessionStorage.js (hunk 4 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/common/sessionStorage.js:1]: than the file name should indicate its part of test.

File: webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js [hunk 10]
```diff
@@ -2,24 +2,22 @@ import React from 'react';
 import { Dropdown, Glyphicon, MenuItem } from 'react-bootstrap';
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js]: nitpick: the filename can now be simply DropDown as its in the directory notifications/drawer?

Long term, I would assume the same component should work for toast notifications too?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/index.js (hunk 7 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/drawer/index.js]: also, @gailsteiger implementation was with H4 vs H3, which one is correct?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js [hunk 11]
```diff
@@ -1,39 +1,50 @@
...
 
 /* eslint-disable camelcase */
-const Notification = ({created_at, seen, text, level, id, actions}) => {
-  const created = moment(created_at);
+const Notification = (
+  {
+    notification: {
+      created_at,
+      seen,
+      text = '-',
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js]: why is this needed? I can't assume a server response with "-" as the text?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js [hunk 12]
```diff
@@ -1,39 +1,50 @@
...
+        onClick={onMarkAsRead.bind(this, id)}>
+        { text }
+    </span>;
 
   return (
     <div className="drawer-pf-notification">
-      <Icon type={level} css="pull-left"></Icon>
-      {actions.links && <NotificationDropdown links={actions.links} id={id} />}
+      <Icon type={level} />
+      <div className="notification-text-container">
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js]: nitpick: Indentation is wrong.

File: webpack/assets/javascripts/react_app/components/notifications/drawer/notificationGroup.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  return (
+    <div className={className}>
+      <div className="panel-heading" onClick={() => onExpand(group)}>
+        <h4 className="panel-title">
+          <a className={isExpanded ? '' : 'collapsed'}>
+            {group}
+          </a>
+        </h4>
+        <span className="panel-counter">
+          {unreadCount} New {unreadCount !== 1 ? 'Events' : 'Event'}
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/notificationGroup.js]: please extract.

File: webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        seen: true,
+        level: 'info',
+        text: 'Hi! This is a notification message',
+        'created_at': '2017-03-14T11:25:07.138-04:00',
+        group: 'React devs2',
+        actions: {
+          links: [
+            {
+              href: 'https://theforeman.org/blog',
+              title: 'Link to blog'
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js:39]: should be external i assume?

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  NOTIFICATIONS_GET_NOTIFICATIONS,
+  NOTIFICATIONS_TOGGLE_DRAWER,
+  NOTIFICATIONS_SET_EXPANDED_GROUP,
+  NOTIFICATIONS_MARK_AS_READ
+} from '../../consts';
+import { notificationsDrawer } from '../../../common/sessionStorage';
+import API from '../../../API';
+
+export const getNotifications = url => dispatch => {
+  API.get(url).then(response => dispatch({
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: in the current implementation there are two addition cases which are missing:
1. when there is a failure, stop pulling notifications (e.g. session timeout, or server 500 error).
2. when you are not in focus, we dont want to pull data from the server when there is no one to consume it.

File: webpack/assets/javascripts/react_app/redux/index.js [hunk 16]
```diff
@@ -9,8 +9,12 @@ if (process.env.NODE_ENV !== 'production') {
   middleware = [...middleware, createLogger()];
 }
 
-export default createStore(
+const _createStore = () => createStore(
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/index.js]: can you explain how is this used? should this be as part of this pr?

**Review by ohadlevy — CHANGES_REQUESTED** [review]
@matanwerbner thanks for your PR.

This is a very large PR, that normally should be split to small chunks (different commits in the same PR) or even multiple PRs.

When you are updating this PR, please indicate via a comment what exactly did you fix, so the reviewer (me) doesnt need to re-read the entire pr - thanks!

I've left some comments inline, please have a look, I'll work on the UI changes of notifications next.

File: webpack/assets/javascripts/foreman_tools.js (hunk 2 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/foreman_tools.js]: great point, i'll look into that.

File: webpack/assets/javascripts/react_app/common/sessionStorage.js (hunk 4 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/common/sessionStorage.js:1]: it's not a part of test, this is a service file for sessionStorage encapsulation.

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/common/sessionStorage.js:1]: but its used only for tests right?

**Comment by ohadlevy** [comment]
when expanding groups, i see the following error:
```js
index.js:120 Uncaught Error: Actions may not have an undefined "type" property. Have you misspelled a constant?
    at Object.performAction (<anonymous>:3:1870)
    at liftAction (<anonymous>:2:27428)
    at dispatch (<anonymous>:2:31406)
    at index.js:104
    at index.js:14
    at bindActionCreators.js:7
    at onClick (notificationGroup.js:21)
```

can you please add a test for it? thanks.

File: webpack/assets/javascripts/react_app/redux/index.js (hunk 16 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/redux/index.js]: removed

File: webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js (hunk 14 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js:39]: there's no test for that currently

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 17]
```diff
@@ -8,12 +8,17 @@ import { notificationsDrawer } from '../../../common/sessionStorage';
...
-    type: NOTIFICATIONS_GET_NOTIFICATIONS,
-    payload: {
-      notifications: response.notifications
-    }
-  }));
+  if (
+    document.visibilityState === 'visible' ||
+    document.visibilityState === 'prerender'
+  ) {
+    API.get(url).then(response => dispatch({
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: also, does this make sure the API calls happens only after the last completed? e.g. it will wait the TIMER interval from when the last successful call.

File: webpack/assets/javascripts/react_app/components/notifications/notifications.consts.js [hunk 18]
```diff
@@ -0,0 +1,4 @@
+export const getNotificationsInterval = 10000;
+
+export const NO_NOTIFICATIONS_MESSAGE = 'No Notifications';
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/notifications.consts.js]: please extract!

**Comment by ohadlevy** [comment]
Please note the visual changes across the old vs new:

NEW

image

OLD
image


no min height ?
Group header looks different size?

File: webpack/assets/javascripts/react_app/components/notifications/index.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+import { getNotificationsInterval } from './notifications.consts';
+
+class notificationContainer extends React.Component {
+  componentDidUpdate() {
+    window.tfm.tools.activateTooltips();
+  }
+
+  componentDidMount() {
+    const { getNotifications, data: { url } } = this.props;
+    const timerId = setInterval(
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/index.js]: IMHO SetInterval is not a good use here, as I would not want to fire API call every 10 seconds, but rather 10s from when the last response was.

This avoid the issue when a server is slow to respond and events just pile up (or multiple API calls for the same URL are pending).

**Comment by ohadlevy** [comment]
@matanwerbner see https://gist.github.com/ohadlevy/ea3e2e17aad226031a9ff9dcdbb6619f for fixes.

File: npm-debug.log [hunk 20]
```diff
@@ -1,45 +0,0 @@
-0 info it worked if it ends with ok
-1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'run', 'test' ]
-2 info using npm@4.1.2
-3 info using node@v7.5.0
-4 verbose run-script [ 'pretest', 'test', 'posttest' ]
```
Comment (ohadlevy) [inline npm-debug.log]: please remove this file

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+    document.visibilityState === 'prerender'
+  ) {
+    API.get(url).then(response => {
+      dispatch({
+        type: NOTIFICATIONS_GET_NOTIFICATIONS,
+        payload: {
+          notifications: response.notifications
+        }
+      });
+      setTimeout(
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: this line should be outside the if statement, otherwise once the condition is not met, future timers will not be executed.

File: webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js (hunk 5 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js]: please ensure that once the target is self, then close the notification drawer when clicking on an action.

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js]: see #4403 for details

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 21 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: you asked for the timeout to begin after the response was received, no?

File: webpack/assets/javascripts/react_app/common/sessionStorage.js (hunk 4 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/common/sessionStorage.js:1]: nope. it's used in redux/actions/notifications/index.js as well

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 21 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: right, but in the case where it was skipped (as the tab was not in focus) it should trigger the same check again, the current code will stop first time the condition is not met.

File: app/assets/stylesheets/patternfly_and_overrides.scss [hunk 22]
```diff
@@ -352,9 +352,4 @@ a.btn-info span{
   z-index: auto !important;
 }
 
-// workaround to patternfly css after removing class navbar-pf from navbar
-// to retain original navbar height
-.navbar-outer .drawer-pf {
-  height: calc(100vh - 46px);
-  top: 26px;
-}
+
```
Comment (tbrisker) [inline app/assets/stylesheets/patternfly_and_overrides.scss]: nitpick: extra white space

File: webpack/assets/javascripts/react_app/common/MountingService.js [hunk 23]
```diff
@@ -1,8 +1,7 @@
 import React from 'react';
 import StatisticsChartsList from '../components/charts/StatisticsChartsList';
 import PowerStatus from '../components/hosts/powerStatus/';
-import NotificationDrawerToggle from '../components/notifications/NotificationDrawerToggle';
-import NotificationDrawer from '../components/notifications/NotificationDrawer';
+import NotificationDrawer from '../components/notifications/';
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/common/MountingService.js]: Isn't this supposed to be NotificationContainer?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+      level,
+      id,
+      actions
+    },
+    onMarkAsRead,
+    onClickedLink
+  }
+) => {
+  const created = new Date(created_at);
+  const title = __('Click to mark as read').toString();
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js]: Isn't this already a string?

File: webpack/assets/javascripts/react_app/components/notifications/notifications.consts.js [hunk 25]
```diff
@@ -0,0 +1,2 @@
...
+export const NOTIFICATIONS_TITLE = 'Notifications';
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/notifications/notifications.consts.js]: These also need extraction

File: webpack/assets/javascripts/react_app/components/notifications/notifications.test.js [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+      return new Promise(resolve => resolve(JSON.parse(serverResponse)));
+    });
+  });
+
+  it('empty state', () => {
+    const store = mockStore(emptyState);
+
+    const box = shallow(<Notifications store={store} />);
+
+    expect(box.render().html()).toEqual(emptyHtml);
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/notifications/notifications.test.js:40]: Would it make sense to check that certain properties of the html are correct instead of comparing the entire html output here and in the following tests? That way if we change some minor part in the html the test won't break.

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+    return null;
+  }).then(
+    () => {
+      setTimeout(
+        () => dispatch(getNotifications(url)),
+        getNotificationsInterval
+      );
+    },
+    // error handling
+    () => {}
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:40]: Would it make sense I think to display some message in the drawer or maybe even change the bell color so the user will have some indication that something failed and what to do about it?

**Review by tbrisker — CHANGES_REQUESTED** [review]
Looks good in general, just a few mostly minor inline comments.

File: webpack/assets/javascripts/react_app/actions/NotificationActions.js (hunk 9 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/actions/NotificationActions.js:10]: added.

File: webpack/assets/javascripts/react_app/components/notifications/drawer/notificationGroup.js [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+        </span>
+      </div>
+      {
+        isExpanded &&
+        <div className="panel-body">
+          {
+            orderBy(notifications, 'created_at', 'desc').map(
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/notificationGroup.js]: aren't they already ordered?

File: webpack/assets/javascripts/react_app/components/notifications/notifications.consts.js (hunk 25 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/notifications.consts.js]: it's extracted in the UI component

File: webpack/assets/javascripts/react_app/components/notifications/notifications.test.js (hunk 26 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/notifications.test.js:40]: indeed, changed.

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 27 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:40]: @ohadlevy your thoughts?
anyway i dont think this is urgent for this PR

File: app/controllers/notification_recipients_controller.rb [hunk 29]
```diff
@@ -7,11 +7,12 @@ class NotificationRecipientsController < Api::V2::BaseController
   def index
     @notifications = NotificationRecipient.
       where(:user_id => User.current.id, :notification_id => Notification.active).
-      order(created_at: :desc).
+      order(:created_at).
```
Comment (tbrisker) [inline app/controllers/notification_recipients_controller.rb]: Why did you revert this?

File: app/controllers/notification_recipients_controller.rb [hunk 30]
```diff
@@ -7,11 +7,12 @@ class NotificationRecipientsController < Api::V2::BaseController
...
       eager_load(:notification, :notification_blueprint)
 
     render :json => {
-      :notifications => @notifications.paginate(paginate_options).map(&:payload)
+      :notifications => @notifications.paginate(paginate_options).map(&:payload),
+      :total => @notifications.count
```
Comment (tbrisker) [inline app/controllers/notification_recipients_controller.rb]: Also this?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/notificationGroup.js [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+          {
+            `${unreadCount} ${unreadCount !== 1 ?
+              __('New Events') :
+              __('New Event')}`
+          }
+        </span>
+      </div>
+      {isExpanded &&
+        <div className="panel-body">
+          {orderBy(notifications, 'created_at', 'desc').map(notification => (
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/notificationGroup.js]: Isn't this sorted already?

**Review by tbrisker — COMMENTED** [review]
Looks like your last commit reverted the ordering change on the server side?

File: app/controllers/notification_recipients_controller.rb (hunk 30 above)
Comment (matanwerbner) [reply app/controllers/notification_recipients_controller.rb]: @ohadlevy said to return the total, since it's breaking the tests, so i reverted the entire file.

**Comment by ohadlevy** [comment]
@tbrisker I've asked @matanwerbner to reduce the server side changes for the total counts as it was breaking tests, since I already implemented the same fix at #4383 I think it make sense to merge that first and rebase?

File: webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+        <OverlayTrigger placement="top" overlay={tooltip}>
+        < span>{ text }</span>
+        </OverlayTrigger>
+    </span>;
+
+  return (
+    <div className="drawer-pf-notification">
+      <Icon type={level} />
+      <div className="notification-text-container">
+      {markup}
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/notification.js]: nitpick: missing indentation.

File: webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js (hunk 14 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js:39]: not sure i understand your comment? ;)

File: webpack/assets/javascripts/react_app/components/notifications/toggleIcon/index.js [hunk 33]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+
+export default ({ hsUnreadMessages, onClick }) => {
+  const iconType = hsUnreadMessages ? 'fa-bell' : 'fa-bell-o';
+
+   return (
+   <a className="nav-item-iconic drawer-pf-trigger-icon" onClick={onClick}>
+        <span className={`fa ${iconType}`} title={__('Notifications')}>
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/toggleIcon/index.js]: do you want to use either the Icon class or the react-bootstrap tooltips here too?

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 27 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:40]: :+1: to not include this in the PR, I'm uncertain what should we display in those cases as well.

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 34]
```diff
@@ -0,0 +1,84 @@
...
+  );
+};
+
+export const onMarkAsRead = (group, id) => dispatch => {
+  dispatch({
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:45]: this fixes http://projects.theforeman.org/issues/18853

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 35]
```diff
@@ -0,0 +1,84 @@
...
+    type: NOTIFICATIONS_TOGGLE_DRAWER,
+    payload: {
+      value: !isDrawerOpened
+    }
+  });
+};
+
+export const onClickedLink = link => (dispatch, getState) => {
+  toggleDrawer()(dispatch, getState);
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:82]: this fixes http://projects.theforeman.org/issues/18678

File: webpack/assets/javascripts/react_app/components/notifications/drawer/index.js [hunk 36]
```diff
@@ -0,0 +1,41 @@
...
+        onClickedLink={onClickedLink}
+        onMarkAsRead={onMarkAsRead}
+        isExpanded={expandedGroup === key}
+        onExpand={onExpandGroup}
+        notifications={notificationGroups[key]}
+      />
+    ));
+  const noNotificationsMessage = (
+    <div id="no-notifications-container">
+      {__('No Notifications')}
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/index.js:27]: this fixes http://projects.theforeman.org/issues/18193

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 27 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:40]: however, I do think that on failure, we should stop pulling.
the same goes for session timeout.

(this was part of the initial implementation).

Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:40]: This is whats happening, as the error handler does not call the get again...

File: webpack/assets/javascripts/react_app/components/notifications/notifications.test.js (hunk 26 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/notifications.test.js:40]: However it does catch small bugs that otherwise go un-noticed, like the time created being different on the local machine then the build machine...

File: webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js (hunk 14 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js:39]: This is fixture file, used for tests. Since there is no test currently about clicking the link,adding external property to the fixture doesnt matter

File: webpack/assets/javascripts/react_app/components/notifications/toggleIcon/index.js (hunk 33 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/react_app/components/notifications/toggleIcon/index.js]: tooltips: absolutly yes.
Icon - not sure, i think we can do a better implementation of it, that supports onClick etc.
we should make an effort to standardize icons in a better, more friendly way in a seperate PR

File: app/controllers/notification_recipients_controller_BACKUP_16002.rb [hunk 37]
```diff
@@ -0,0 +1,41 @@
+class NotificationRecipientsController < Api::V2::BaseController
+  include Foreman::Controller::Parameters::NotificationRecipient
+
+  before_action :require_login
+  before_action :find_resource, :only => [:update, :destroy]
+
+  def index
+<<<<<<< upstream/develop
```
[the remark this answers is not in the document]
Comment (ohadlevy) [reply app/controllers/notification_recipients_controller_BACKUP_16002.rb]: please remove all of the rebase files.

File: webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js [hunk 38]
```diff
@@ -0,0 +1,25 @@
...
+    <Dropdown pullRight className="dropdown-kebab-pf" id={id}>
+      <Dropdown.Toggle noCaret bsStyle="link">
+        <Glyphicon bsClass="fa" glyph="ellipsis-v" />
+      </Dropdown.Toggle>
+      <Dropdown.Menu>
+        {
+          links.map((link, i) =>
+            <MenuItem key={i}
+                      id={ `notification-kebab-${i}` }
+                      onClick={ onClickedLink.bind(this, link) }>
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js:15]: extra ">" at the end of the line?

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/notifications/drawer/NotificationDropdown.js:15]: oh i see, the indentation confused me.

**Comment by dLobatog** [comment]
Why is this 1.15 critical? @ohadlevy ?

**Comment by ohadlevy** [comment]
@dLobatog I dont want to release 1.15 with flux if i dont have to.

**Review by ohadlevy — APPROVED** [review]

**Comment by ohadlevy** [comment]
thanks @matanwerbner!
